add golangci-linter, simplify travis.yml, delint some small stuff

aerth opened the pull request:
Reorder OS tests, so that we run the Windows tests first.

Simplify travis.yml, make all OS run testdata/travis.bash (where it's easier to switch on conditions)

Simplify error check in nss.go

Miscellaneous whitespace-related delinting, left some non-trivial things for another time/place

Since Go module support was added in Go 1.11... the Go 1.10 tests we weren't using the pull request's source code (we were fetching master branch of this repo). To workaround this, we copy the source into `$GOPATH/src/github.com/namecoin/certinject` to avoid downloading the master branch. This only happens in Go 1.10 test run. For go 1.11 and 1.12, we set GO111MODULE to activate go module support. (its active by default in newer versions)

Add a **lint** stage, which runs golangci-linter. It's allowed to fail for now, but the travis build logs are helpful.

aerth committed:
tests: add golangci-linter, simplify travis.yml, workaround for <go1.11

aerth committed:
delint: don't export Log (xlog.Site), instead provide SetLogLevel() func

aerth committed:
delint: misc trivial delinting

suppress gosec sha1 rules

JeremyRand commented on .travis.yml:
```diff
@@ -19,14 +19,34 @@ notifications:
     # GITHUB_TOKEN for automatic releases
     - secure: "at1oJs7ib7glx3W+zk+OkT041LdknVXirIhN403CIihVUrlOhODY7yCTgvF4Rk0jYBJiT35Q2qxpgfWF2qGnsNsQmjG3ydDWQDCepDc/CgXfLyoiSTJK5vTK72dYWTVsBTycXbj1CbSy2X2ah/KWjc4RcgZ67ER7mDpRU5nFeow="
-    # Set this to the Go version to use for releases (must appear in version list above).
+      # Set this to the Go version to use for releases (must appear in version list above).
```
Why is the whitespace added here?  AFAICT it flows better without the extra whitespace.

JeremyRand commented on .travis.yml:
```diff
@@ -19,14 +19,34 @@ notifications:
+      # Set this to the Go version to use for releases (must appear in version list above).
     - RELEASE_GO_VERSION="1.x"
+      # Go versions before go1.13.x will grab master branch unless we use GO111MODULE=on
+      # (go1.10.x still needs the above workaround)
```
Which "above workaround" is being referred to here?

aerth replied:
Nice. This was from before I moved the travis script from travis.yml into testdata/travis.bash.

I updated this comment to point to testdata/travis.bash for the workaround (in 2ffe3be) and added a bit more about it in travis.bash (in 7a9a9a1)

JeremyRand commented on cryptoapi_windows.go:
```diff
@@ -1,7 +1,7 @@
 
 import (
-	"crypto/sha1"
+	"crypto/sha1" //nolint:gosec
```
I assume this is specifically an attempt to suppress `gosec` rule G505?  If so, please use `// #nosec G505` instead, so that we don't suppress other `gosec` warnings.

JeremyRand commented on cryptoapi_windows.go:
```diff
@@ -138,7 +138,7 @@ func injectCertCryptoApi(derBytes []byte) {
 	// This is probably a Bad Thing (TM) since SHA-1 is weak.
 	// However, that's Microsoft's problem to fix, not ours.
-	fingerprint := sha1.Sum(derBytes)
+	fingerprint := sha1.Sum(derBytes) //nolint:gosec
```
As above, please suppress a specific `gosec` rule rather than the entire `gosec` linter.

aerth committed:
travis.yml: fix whitespace

aerth committed:
travis.bash: document and simplify workaround for no Go module support

also avoiding the -u flag which can change dependencies versions

(the -u flag isn't needed because the travis build environment shouldn't
have any go libraries installed anyways)

basically, it should be doing exactly what we expect it to do
now.

JeremyRand commented on testdata/golangci-linter.bash:
```diff
@@ -0,0 +1,14 @@
+  --color always \
+  --disable gochecknoglobals,gomnd \
+  -v \
+  ./...
```
* Maybe add a `GOOS=linux` CI run as well?  There are a few source files that aren't built for Windows.
* Maybe add an extra CI run with `--new-from-rev=HEAD~`, which isn't allowed to fail.
* Maybe add an extra CI run with only the linters that are currently known to pass, which isn't allowed to fail.

JeremyRand commented on testdata/travis.bash:
```diff
@@ -1,6 +1,26 @@
+#!/bin/bash
 set -ex
+
+# workaround for go1.10's no module support, we copy this run's source code into namecoin/certinject
```
Can you elaborate this comment a bit?  Not totally clear to me at first glance why `$TRAVIS_REPO_SLUG` is relevant.

JeremyRand commented:
utACK 82f3198fc04710fea48daccbbb75aa19da402948 except for the inline review I posted above.

aerth committed:
linter: run GOOS=windows and GOOS=linux golangci-linter

aerth committed:
travis.yml: add linter stage for --new-from-rev=HEAD~

JeremyRand commented:
Still remaining from my review:

* Maybe add an extra CI run with only the linters that are currently known to pass, which isn't allowed to fail.

utACK 13b62d35f55fe30f0537c471798dc48342a38a54; I'm fine with deferring the remaining review item to a separate issue/PR.
